ruby/ql/lib/codeql/ruby/frameworks/ActionDispatch.qll

```diff
@@ -0,0 +1,902 @@
+    RouteBlock parent;
+    string resource;
+    string action;
+    string httpMethod;
+    string pathComponent;
+
+    SingularResourceRoute() {
+      this = TResourceRoute(parent, method, action) and
+      resource = method.getArgument(0).(StringlikeLiteral).getValueText() and
+      isDefaultSingularResourceRoute(resource, httpMethod, pathComponent, action)
+    }
+
+    override string getAPrimaryQlClass() { result = "SingularResourceRoute" }
+
+    override RouteBlock getParentBlock() { result = parent }
+
+    override string getLastPathComponent() { result = pathComponent }
+
+    override string getLastControllerComponent() { result = method.getArgument(0).getValueText() }
+
+    override string getAction() { result = action }
+
+    override string getHTTPMethod() { result = httpMethod }
+  }
+
+  /**
+   * A route generated by a call to `match`.
+   * This is a lower level primitive that powers `get`, `post` etc.
+   * The first argument can be a path or a (path, controller-action) pair.
+   * The controller, action and HTTP method can be specified with the
+   * `controller:`, `action:` and `via:` keyword arguments, respectively.
+   * ```ruby
+   * match 'photos/:id' => 'photos#show', via: :get
+   * match 'photos/:id', to: 'photos#show', via: :get
+   * match 'photos/:id', controller: 'photos', action: 'show', via: :get
+   * ```
+   */
+  class MatchRoute extends Route, TMatchRoute {
+    private RouteBlock parent;
+
+    MatchRoute() { this = TMatchRoute(parent, method) }
+
+    override string getAPrimaryQlClass() { result = "MatchRoute" }
+
+    override RouteBlock getParentBlock() { result = parent }
+
+    override string getLastPathComponent() {
+      result = method.getArgument(0).(StringlikeLiteral).getValueText() or
+      result = method.getArgument(0).(Pair).getKey().getValueText()
+    }
+
+    override string getLastControllerComponent() {
+      result = extractController(method.getKeywordArgument("to").getValueText()) or
+      result = method.getKeywordArgument("controller").getValueText() or
+      result = extractController(method.getArgument(0).(Pair).getValue().getValueText())
+    }
+
+    override string getHTTPMethod() {
+      result = method.getKeywordArgument("via").(StringlikeLiteral).getValueText() or
+      result = method.getKeywordArgument("via").(ArrayLiteral).getElement(_).getValueText()
+    }
+
+    override string getAction() {
+      result = extractAction(method.getKeywordArgument("to").getValueText()) or
+      result = method.getKeywordArgument("action").getValueText() or
+      result = extractAction(method.getArgument(0).(Pair).getValue().getValueText())
+    }
+  }
+
+  /**
+   * Several routing methods support the keyword arguments `only:` and `except:`.
+   * - `only:` restricts the set of actions to just those in the argument.
+   * - `except:` removes the given actions from the set.
+   */
+  bindingset[action]
+  predicate applyActionFilters(MethodCall m, string action) {
```

Comment:
I think a lot of these helper predicates can/should be `private`.

Reply:
Good point. I've made everything private except `Route` and `underscore`, which are both used in `ActionController` to tie routes to action handlers.